kernel/src/process_standard.rs

```diff
@@ -1385,14 +1386,46 @@ impl<C: 'static + Chip> ProcessStandard<'_, C> {
                         "Process not enabled flash={:#010X}-{:#010X} process={:?}",
                         app_flash.as_ptr() as usize,
                         app_flash.as_ptr() as usize + app_flash.len() - 1,
-                        process_name
+                        process_name.unwrap_or("(no name)")
                     );
                 }
             }
             // Return no process and the full memory slice we were given.
             return Ok((None, remaining_memory));
         }
 
+        if let Some((major, minor)) = tbf_header.get_kernel_version() {
+            // If the `KernelVersion` header is present, we read the requested kernel version and compare it to
+            // the running kernel version.
+            if crate::MAJOR != major {
+                // If the kernel major version is different, we prevent the process from being loaded.
+                if config::CONFIG.debug_load_processes {
+                    debug!("WARN process {:?} not loaded as it requires kernel version {} (running kernel {}.{})", process_name.unwrap_or("(no name)"), major, crate::MAJOR, crate::MINOR);
+                }
+                return Ok((None, remaining_memory));
+            } else if crate::MINOR < minor {
+                // If the kernel major version is the same, we compare the kernel minor version. The current
+                // running kernel minor version has to be greater or equal to the one that the process
+                // has requested. If not, we prevent the process from loading.
+                if config::CONFIG.debug_load_processes {
+                    debug!("WARN process {:?} not loaded as it requires kernel version at least {}.{}, (running kernel {}.{})", process_name.unwrap_or("(no name)"), major, minor, crate::MAJOR, crate::MINOR);
+                }
+                return Ok((None, remaining_memory));
+            }
+        } else {
+            if enforce_kernel_version {
+                // If enforing the kernel version is requested, and the `KernelVersion` header is not present,
+                // we prevent the process from loading.
+                if config::CONFIG.debug_load_processes {
+                    debug!(
+                            "WARN process {:?} not loaded as it has no kernel version header, please upgrade to elf2tab >= 0.8.0",
+                            process_name.unwrap_or ("(no name")
+                        );
+                }
+                return Ok((None, remaining_memory));
```

Comment:
I think we should return an error here. The caller asked to enforce kernel version, and the check failed.

Reply:
My understanding is that if this returns an error, `load_processes_advanced` will stop loading processes, not just skip this process.

```rust
  ProcessStandard::create(
                    kernel,
                    chip,
                    entry_flash,
                    header_length as usize,
                    version,
                    remaining_memory,
                    fault_policy,
                    enforce_kernel_version,
                    i,
            )?
```

Reply:
That's correct. But without an error this will silently fail, and it's hard to differentiate an app that just doesn't do anything from one that was never actually loaded. Presumably no user who wants the kernel version compatibility checked wants to just ignore those errors.

Reply:
Agree, but on the other hand this would make "some apps" (all apps after the one that cannot be loaded) silently fail.

What about always showing this warnings if enforce kernel version is true?

Reply:
What is the use case for knowing that you have an incompatible app but continuing to run the others? I suspect that the most likely case is the user only has one app (or if there are multiple they are all incompatible) and they are going to fix it immediately.

Reply:
The user might have loaded one of the apps from a tab using tockloader. The issue is that it will load all apps up to the one that does not meet the requirements or has no header. In case we do not want to load any app, maybe it would be a good idea to panic as it is not deterministic which apps will run.

Reply:
Most boards do not panic, but do print a warning/error when load_processes fails. I've heard anecdotally that the checks for an app that does not meet ram/flash position requirements which cause load_processes to fail have worked well.

Reply:
I got it now :)  

I just saw that most boards report a debug message upon an error.